README.rst
karl.external_link_ticket
=========================

A way to wrap external links with a ticket that can be validated by the
external system, thus acting as a psuedo SSO.

Explanation
===========

In KARL, we have some links in the OSF home page that go to external
resources in the Business Center. However, in KARL, we do some wrapping
of those external links:

- In KARL, the URL looks like:

 https://karl.soros.org/wrap_external_link?external_url=http://osi-ny.soros.org/troubletkt/step1.cfm

- The user clicks on that and goes to /wrap_external_link in KARL

- We then generate a "ticket" and store it in the KARL database

- The ticket has these fields: email, external_url, remote_addr,
  created, and used

- The ticket value is stored with a key that is a random 32 bit value

- We then redirect the user to the actual URL, adding:

   ?karl_authentication_ticket=32bitkey

 ...at the end.

In the external system, they then want to validate that KARL issued
the value in the 32bitkey. That system:

- Sends a request to such as:

   https://karl.soros.org/authenticate_ticket

 ..with 'ticket', 'external_url', and 'remote_addr' in the data.

- We then "validate" the ticket:

 * Does it exist?

 * Has it been used?

 * Is it over a minute old?

 * Is it for the external URL that it was issued for?

- If not valid, we respond with an XML message indicating the failure

- If it is valid, we respond with a success message


CHANGES.txt
karl.external_link_ticket package Changelog
===========================================

3.5 (2013-11-22)
----------------

- Help with the ZenDesk API changes by sending first and last name in
  the XML response to ``authenticate_ticket`` POST requests
  (LP #1251677)

3.4 (2011-10-27)
----------------

- Pyramid updates

3.3 (2011-10-27)
----------------

- Bring up to date with bfg 1.1a5 (merge formish branch).

3.2 (2009-07-23)
----------------

- Update index url.

3.1 (2009-06-24)
----------------

- Disabled checking of REMOTE_HOST

3.0 (2009-06-15)
----------------

- Just updated version number and tagging scheme.

osi-production-20090611 (2009-06-11)
------------------------------------

- No changes

osi-production-20090610 (2009-06-10)
------------------------------------

- No changes

osi-production-20090608 (2009-06-08)
------------------------------------

- No changes in this package.

osi-production-20090603 (2009-06-03)
------------------------------------

- Changed authenticate_ticket to be consistent with the existing client-side
  code by accepting the key as the variable ticket.

- Made authenticte_ticket view accessible to anonymous users.

osi-production-20090531 (2009-05-31)
------------------------------------

- No changes to this package.

osi-production-20090528 (2009-05-28)
------------------------------------

- Initial version rolled out to replace KARL2.



karl/external_link_ticket/scripts/tests.py
# Copyright (C) 2008-2009 Open Society Institute
#               Thomas Moroz: tmoroz@sorosny.org
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License Version 2 as published
# by the Free Software Foundation.  You may not use, modify or distribute
# this program under any other version of the GNU General Public License.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.

import unittest
from zope.testing.cleanup import cleanUp
from repoze.bfg import testing
from karl import testing as karltesting
from repoze.lemonade.testing import registerContentFactory
from karl.models.interfaces import IProfile

class SyncStaffRunnerTests(unittest.TestCase):

    def setUp(self):
        cleanUp()

    def tearDown(self):
        cleanUp()

    def _getTargetClass(self):
        from karl.peopledir.scripts.sync_osi_staff import SyncStaff
        return SyncStaff

    def _makeOne(self, args=(), test_file='test1.xml'):
        # Load some default data from the sampledata dir
        from os.path import join
        from repoze.bfg.path import package_path
        from karl.peopledir import scripts
        scripts_path = package_path(scripts)

        full_sync_file = join(scripts_path, 'sampledata', test_file)
        argv = ['testing'] + list(args) + [full_sync_file]
        sync_staff = self._getTargetClass()(argv, testing=True)
        return sync_staff

    def test___init___defaults(self):
        sync_staff = self._makeOne(test_file='test1.xml')
        self.assertEqual(sync_staff.zodb_uri, 'zeo://localhost:8884')
        self.assertEqual(sync_staff.sync_file[-9:], 'test1.xml')

    def test_wellformed(self):
        sync_staff = self._makeOne(test_file='test1.xml')
        sync_staff.loadXML()
        self.assertEqual(sync_staff.xml_root.tag, 'users')

    def test_not_wellformed(self):
         from lxml.etree import XMLSyntaxError
         sync_staff = self._makeOne(test_file='test2.xml')
         self.assertRaises(XMLSyntaxError, sync_staff.loadXML)

    def test_wellformed_invalid(self):
        from lxml.etree import DocumentInvalid
        sync_staff = self._makeOne(test_file='test3.xml')
        sync_staff.validate = True
        self.assertRaises(DocumentInvalid, sync_staff.loadXML)

# Skip this because we aren't refactoring deeply inside sync.py to
# eliminate passing around XML.

#     def test_normalize_users(self):
#         sync_staff = self._makeOne(test_file='test4.xml')
#         sync_staff.loadXML()
#         users = sync_staff.normalize_users()
#         self.assertEqual(users, _sample_data_test4)

    def test_add_new_user(self):
        # registration
        registerContentFactory(karltesting.DummyProfile, IProfile)

        sync_staff = self._makeOne(test_file='test4.xml')
        sync_staff.loadXML()

        # Fake out the wiring up of users etc.
        sync_staff.site = testing.DummyModel()
        sync_staff.profiles = testing.DummyModel()
        sync_staff.users = karltesting.DummyUsers()

        user1 = sync_staff.xml_root.find('user')
        username = user1.findtext('username')
        profile = sync_staff.add_new_user(user1)

        # Check users, groups, and profiles to see if they are there
        by_login = sync_staff.users._by_login
        self.assertEqual(by_login.keys(), [username])
        g = ['group.KarlAffiliate', 'group.BudapestAdmins',
            'group.BaltimoreAdmins', 'group.BrusselsHRAdmins',
            'group.BrusselsAdmins', 'group.BrusselsOMAdmins']
        user = by_login['CpeobsL']
        self.assertEqual(user['groups'], g)

        self.assertEqual(sync_staff.profiles.keys(), ['CpeobsL'])
        profile = sync_staff.profiles['CpeobsL']
        self.assertEqual(profile.firstname, 'Lsjtaujob')
        self.assertEqual(profile.lastname, 'Twfe')
        self.assertEqual(profile.email, 'twfel@dfv.iv')
        self.assertEqual(profile.location, 'Budapest')
        self.assertEqual(profile.country, 'Hungary')
        self.assertEqual(profile.languages,
                         'Hungarian English German French')

_sample_data_test4 = [
            {'username': 'CpeobsL',
             'firstname': 'Somefirst',
             'extension': None, 'lastname':
             'TheLastname',
             'phone': None,
             'organization': 'Major University',
             'position': 'Head of Mega Office',
             'email': 'xx@yy.hu',
             },
            {'username': 'another',
             'firstname': 'Linux',
             'extension': None,
             'lastname': 'Pauling',
             'phone': None,
             'organization': 'Medical World',
             'position': 'Research Director',
             'email': 'linux@pauling.hu',
             },
            ]


karl/external_link_ticket/views/tests/__init__.py
# Copyright (C) 2008-2009 Open Society Institute
#               Thomas Moroz: tmoroz@sorosny.org
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License Version 2 as published
# by the Free Software Foundation.  You may not use, modify or distribute
# this program under any other version of the GNU General Public License.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
# 
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.

# karl.peopledir views tests


karl/external_link_ticket/views/tests/test_utils.py
import unittest

from pyramid import testing
from karl.external_link_ticket.testing import DummyTicketContext

class TestFindLinkTickets(unittest.TestCase):
    def setUp(self):
        testing.setUp()

    def tearDown(self):
        testing.tearDown()

    def _callFUT(self, context):
        from karl.external_link_ticket.views.utils import find_link_tickets
        return find_link_tickets(context)

    def test_valid_created(self):
        context = testing.DummyModel()
        result = self._callFUT(context)
        self.assertEqual(result.__name__, 'link_tickets')

    def test_valid_existing(self):
        context = testing.DummyModel()
        link_tickets = testing.DummyModel(title='existing link tickets')
        context['link_tickets'] = link_tickets
        result = self._callFUT(context)
        self.assertEqual(result.__name__, 'link_tickets')
        self.assertEqual(result.title, 'existing link tickets')


class TestGenerateTicket(unittest.TestCase):
    def setUp(self):
        testing.setUp()

    def tearDown(self):
        testing.tearDown()

    def _callFUT(self, context, firstname, lastname, email, remote_addr,
                 external_url):
        from karl.external_link_ticket.views.utils import generate_ticket
        return generate_ticket(context, firstname, lastname,
                               email, remote_addr, external_url)

    def test_valid(self):
        context = testing.DummyModel()
        key = self._callFUT(context, 'firstname', 'lastname',
                            'test@example.com', '192.168.1.1', 'http://example.com')
        self.assertEqual(len(key), 32)


class TestWriteTicket(unittest.TestCase):
    def setUp(self):
        testing.setUp()

    def tearDown(self):
        testing.tearDown()

    def _callFUT(self, context, ticket, key):
        from karl.external_link_ticket.views.utils import write_ticket
        return write_ticket(context, ticket, key)

    def test_it(self):
        ticket = {
            'email': 'test@example.com',
            'remote_addr': '192.168.1.1', 
            'external_url': 'http://example.com',
        }
        key = '123456'
        context = testing.DummyModel()
        result = self._callFUT(context, ticket, key)
        ticket = context['link_tickets'][key]
        self.assertEqual(ticket.email, 'test@example.com')
        self.assertEqual(ticket.remote_addr, '192.168.1.1')
        self.assertEqual(ticket.external_url, 'http://example.com')
        self.failIf(ticket.created == None)
        self.failIf(getattr(ticket, 'used', 'no attribute') == 'no attribute')
        self.assertEqual(ticket.used, None)


class TestGetTicket(unittest.TestCase):
    def setUp(self):
        testing.setUp()

    def tearDown(self):
        testing.tearDown()

    def _callFUT(self, context, key):
        from karl.external_link_ticket.views.utils import get_ticket
        return get_ticket(context, key)

    def test_nonexisting(self):
        context = testing.DummyModel()
        context['link_tickets'] = testing.DummyModel()
        key = '123456'
        result = self._callFUT(context, key)
        self.assertEqual(result, None)
    
    def test_valid(self):
        from datetime import datetime
        created = datetime.now()
        context = DummyTicketContext(**{'created': created,
                                        'first_name': 'First',
                                        'last_name': 'Last'})

        key = '123456'
        result = self._callFUT(context, key)
        self.assertEqual(result['email'], 'test@example.com')
        self.assertEqual(result['created'], created)
        self.assertEqual(result['remote_addr'], '192.168.1.1')
        self.assertEqual(result['external_url'], 'http://example.com')
        self.assertEqual(result['used'], None)


class TestGetTicketAge(unittest.TestCase):
    def setUp(self):
        testing.setUp()

    def tearDown(self):
        testing.tearDown()

    def _callFUT(self, ticket):
        from karl.external_link_ticket.views.utils import get_ticket_age
        return get_ticket_age(ticket)

    def test_valid(self):
        from datetime import datetime
        ticket = {'created': datetime.now()}
        result = self._callFUT(ticket)
        self.assertEqual(result, 0)


class TestValidateTicket(unittest.TestCase):
    def setUp(self):
        testing.setUp()

    def tearDown(self):
        testing.tearDown()

    def _callFUT(self, ticket, remote_addr, external_url):
        from karl.external_link_ticket.views.utils import validate_ticket
        return validate_ticket(ticket, remote_addr, external_url)

    def test_invalid_expired(self):
        from datetime import datetime
        created = datetime(1999,1,1)
        remote_addr = '192.168.1.1'
        external_url = 'http://example.com'
        ticket = {
            'email': 'test@example.com',
            'remote_addr': remote_addr,
            'external_url': external_url,
            'created': created,
            'used': None,
        }
        result = self._callFUT(ticket, remote_addr, external_url)
        self.assertEqual(result['status'], 'FAIL')
        self.assertEqual(result['message'], 'The ticket is expired')

    def test_invalid_used(self):
        from datetime import datetime
        used = datetime(1999,1,1)
        remote_addr = '192.168.1.1'
        external_url = 'http://example.com'
        ticket = {
            'email': 'test@example.com',
            'remote_addr': remote_addr,
            'external_url': external_url,
            'created': datetime.now(),
            'used': used,
        }
        result = self._callFUT(ticket, remote_addr, external_url)
        self.assertEqual(result['status'], 'FAIL')
        self.assertEqual(result['message'], 'The ticket was already used')

    # The remote_addr check is commented out in code since it caused issues with VPN'ed connections
    #def test_invalid_remote_addr(self):
    #    from datetime import datetime
    #    created = datetime.now()
    #    external_url = 'http://example.com'
    #    ticket = {
    #        'email': 'test@example.com',
    #        'remote_addr': '192.168.1.1',
    #        'external_url': external_url,
    #        'created': created,
    #        'used': None,
    #    }
    #    result = self._callFUT(ticket, '192.168.1.2', external_url)
    #    self.assertEqual(result['status'], 'FAIL')
    #    self.assertEqual(result['message'], 'The ticket was generated for a different IP address')

    def test_invalid_external_url(self):
        from datetime import datetime
        created = datetime.now()
        remote_addr = '192.168.1.1'
        ticket = {
            'email': 'test@example.com',
            'remote_addr': remote_addr,
            'external_url': 'http://example.com',
            'created': created,
            'used': None,
        }
        result = self._callFUT(ticket, remote_addr, 'http://example.com/invalid')
        self.assertEqual(result['status'], 'FAIL')
        self.assertEqual(result['message'], 'The ticket was generated for a different external_url')
        

    def test_valid(self):
        from datetime import datetime
        created = datetime.now()
        email = 'test@example.com'
        remote_addr = '192.168.1.1'
        external_url = 'http://example.com'
        ticket = {
            'first_name': 'First',
            'last_name': 'Last',
            'email': 'test@example.com',
            'remote_addr': remote_addr,
            'external_url': external_url,
            'created': created,
            'used': None,
        }
        result = self._callFUT(ticket, remote_addr, external_url)
        self.assertEqual(result['status'], 'SUCCESS')
        self.assertEqual(result['email'], email)


class TestExpireTicket(unittest.TestCase):
    def setUp(self):
        testing.setUp()

    def tearDown(self):
        testing.tearDown()

    def _callFUT(self, context, key):
        from karl.external_link_ticket.views.utils import expire_ticket
        return expire_ticket(context, key)

    def test_nonexisting(self):
        context = testing.DummyModel()
        result = self._callFUT(context, '123456')
        self.assertEqual(result, False)

    def test_valid(self):
        context = DummyTicketContext()
        result = self._callFUT(context, '123456')
        self.assertEqual(result, True)
        self.assertEqual(context['link_tickets'].get('123456',None), None)


karl/external_link_ticket/views/ticket.py
# Copyright (C) 2008-2009 Open Society Institute
#               Thomas Moroz: tmoroz@sorosny.org
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License Version 2 as published
# by the Free Software Foundation.  You may not use, modify or distribute
# this program under any other version of the GNU General Public License.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.


from webob.exc import HTTPFound
from webob import Response
from pyramid.exceptions import Forbidden
from pyramid.security import authenticated_userid
from karl.utils import find_profiles
from karl.external_link_ticket.views.utils import generate_ticket
from karl.external_link_ticket.views.utils import get_ticket
from karl.external_link_ticket.views.utils import validate_ticket
from karl.external_link_ticket.views.utils import expire_ticket

success_xml_template = """<?xml version="1.0"?>
<AuthenticateTicketResponse>
    <Status>%(status)s</Status>
    <Email>%(email)s</Email>
    <FirstName>%(first_name)s</FirstName>
    <LastName>%(last_name)s</LastName>
</AuthenticateTicketResponse>
"""

fail_xml_template = """<?xml version="1.0"?>
<AuthenticateTicketResponse>
    <Status>%(status)s</Status>
    <Message>%(message)s</Message>
</AuthenticateTicketResponse>
"""

def wrap_external_link_view(context, request):
    external_url = request.params.get('external_url', None)
    remote_addr = request.environ.get('REMOTE_ADDR', None)

    if not external_url:
        raise ValueError, 'No external_url provided'

    profiles = find_profiles(context)
    userid = authenticated_userid(request)
    if not userid:
        raise Forbidden('You are not logged in')
    profile = profiles.get(userid)
    if not profile:
        raise Forbidden('No profile found for user %s' % userid)

    first_name = profile.firstname
    last_name = profile.lastname
    key = generate_ticket(context, first_name, last_name, profile.email,
                          remote_addr, external_url)

    if external_url.find('?') != -1:
        location = '%s&karl_authentication_ticket=%s' % (external_url, key)
    else:
        location = '%s?karl_authentication_ticket=%s' % (external_url, key)
    return HTTPFound(location = location)


def authenticate_ticket_view(context, request):
    key = request.params.get('ticket', None)
    external_url = request.params.get('external_url', None)
    remote_addr = request.params.get('remote_addr', None)

    error = None
    if not key:
        error = {'status': 'FAIL', 'message': 'No key provided'}
    if not external_url:
        error = {'status': 'FAIL', 'message': 'No external_url provided'}
    if not remote_addr:
        error = {'status': 'FAIL', 'message': 'No remote_addr provided'}
    if error:
        xml_response = fail_xml_template % error
        return Response(xml_response)

    ticket = get_ticket(context, key)
    if not ticket:
        error = {'status': 'FAIL', 'message': 'No ticket matching key was found'}
        xml_response = fail_xml_template % error
        return Response(xml_response)

    validation_result = validate_ticket(ticket, remote_addr, external_url)
    if validation_result['status'] == 'SUCCESS':
        expire_ticket(context, key)
        xml_response = success_xml_template % validation_result
    else: 
        xml_response = fail_xml_template % validation_result
     
    return Response(xml_response)


karl/external_link_ticket/scripts/__init__.py
# Copyright (C) 2008-2009 Open Society Institute
#               Thomas Moroz: tmoroz@sorosny.org
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License Version 2 as published
# by the Free Software Foundation.  You may not use, modify or distribute
# this program under any other version of the GNU General Public License.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.



karl/external_link_ticket/models/interfaces.py
# Copyright (C) 2008-2009 Open Society Institute
#               Thomas Moroz: tmoroz@sorosny.org
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License Version 2 as published
# by the Free Software Foundation.  You may not use, modify or distribute
# this program under any other version of the GNU General Public License.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.


from zope.interface import Attribute
from zope.interface import Interface

class ILinkTickets(Interface):
    pass

class ILinkTicket(Interface):
    email = Attribute(u'The email address of the user the ticket was generated for')
    external_url = Attribute(u'The external url the ticket was issued for')
    remote_addr = Attribute(u'The ip address of the user requesting the ticket')
    created = Attribute(u'The date/time the ticket was created')
    used = Attribute(u'The date/time the ticket was used')


karl/external_link_ticket/views/utils.py
# Copyright (C) 2008-2009 Open Society Institute
#               Thomas Moroz: tmoroz@sorosny.org
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License Version 2 as published
# by the Free Software Foundation.  You may not use, modify or distribute
# this program under any other version of the GNU General Public License.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.


from datetime import datetime
import random

try:
    from hashlib import md5
except ImportError:
    from md5 import new as md5

from pyramid.security import authenticated_userid
from karl.utils import find_profiles
from karl.utils import find_site
from karl.external_link_ticket.models.ticket import LinkTickets
from karl.external_link_ticket.models.ticket import LinkTicket

success_xml_template = """<?xml version="1.0"?>
<AuthenticateTicketResponse>
    <Status>%(status)s</Status>
    <Email>%(email)s</Email>
</AuthenticateTicketResponse>
"""

fail_xml_template = """<?xml version="1.0"?>
<AuthenticateTicketResponse>
    <Status>%(status)s</Status>
    <Message>%(message)s</Message>
</AuthenticateTicketResponse>
"""

# the number of seconds a ticket should stay active
ticket_ttl = 60


def find_link_tickets(context):
    site = find_site(context)
    link_tickets = site.get('link_tickets', None)
    if not link_tickets:
        link_tickets = LinkTickets()
        site['link_tickets'] = link_tickets
    return link_tickets


def generate_ticket(context, first_name, last_name, email, remote_addr,
                    external_url):
    # generate a random 32 character string
    key = md5(str(random.random())).hexdigest()

    ticket = dict(
        first_name=first_name,
        last_name=last_name,
        email=email,
        remote_addr=remote_addr,
        external_url=external_url,
    )
    write_ticket(context, ticket, key)
    return key


def write_ticket(context, ticket, key):
    link_tickets = find_link_tickets(context)
    link_ticket = LinkTicket(**ticket)
    link_tickets[key] = link_ticket
    return link_ticket


#def __call__(self):
#    external_url = self.request.get('external_url')
#    if not external_url:
#        return 'ERROR: No external_url was passed'
#    get_var_prefix = '?'
#    if external_url.find('?') != -1:
#        get_var_prefix = '&'
#    key = self.generate_ticket(external_url)
#    # redirect user to external_url passing key
#    return self.request.RESPONSE.redirect('%s%skarl_authentication_ticket=%s' % (external_url, get_var_prefix, key))

def get_ticket(context, key):
    link_tickets = find_link_tickets(context)
    link_ticket = link_tickets.get(key, None)
    if not link_ticket:
        return None

    ticket = dict(
        email=link_ticket.email,
        remote_addr=link_ticket.remote_addr,
        external_url=link_ticket.external_url,
        created=link_ticket.created,
        used=link_ticket.used,
        first_name=link_ticket.first_name,
        last_name=link_ticket.last_name,
    )
    return ticket


def get_ticket_age(ticket):
    age = datetime.now() - ticket['created']
    return age.seconds


def validate_ticket(ticket, remote_addr, external_url):
    # Verify that the ticket exists
    if not ticket:
        return {'status': 'FAIL',
                'message': 'No ticket found with key %s' % key}

    # verify the ticket has not been used
    if ticket['used']:
        return {'status': 'FAIL',
                'message': 'The ticket was already used'}

    # verify that the ticket is not more than a minute old
    if get_ticket_age(ticket) > ticket_ttl:
        return {'status': 'FAIL', 'message': 'The ticket is expired'}

    # verify that the ticket was for the correct external_url
    if ticket['external_url'] != external_url:
        return {'status': 'FAIL',
                'message': 'The ticket was generated for a different external_url'}

    # FIXME - commented out for now because VPN'ed users report two different IP's causing a failure
    # verify that the remote_addr is correct
    #remote_addr = remote_addr
    #if remote_addr != ticket['remote_addr']:
    #    return {'status': 'FAIL', 'message': 'The ticket was generated for a different IP address'}

    return {'status': 'SUCCESS', 'email': ticket['email'],
            'first_name': ticket['first_name'],
            'last_name': ticket['last_name']}


def expire_ticket(context, key):
    link_tickets = find_link_tickets(context)
    if key in link_tickets.keys():
        del (link_tickets[key])
        return True
    return False


#    def __call__(self):
#        response_values = self.validate_ticket()
#
#        # render response values into xml
#        if response_values['status'] == 'SUCCESS':
#            xml = success_xml_template % response_values
#        else:
#            xml = fail_xml_template % response_values
#
#        return xml
